### PR TITLE
Refactored ZestBasicRunner not to depend on HttpClient.

### PR DESCRIPTION
**Pull request by denniskniep**
Instead use an interface.

### CODE REVIEW

File: src/main/java/org/mozilla/zest/core/v1/ZestCookie.java [hunk 1]
```diff
@@ -0,0 +1,73 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+package org.mozilla.zest.core.v1;
+
+import java.util.Date;
+
+public class ZestCookie {
```
Comment (thc202) [inline src/main/java/org/mozilla/zest/core/v1/ZestCookie.java]: Should have a since JavaDoc tag, makes life easier for maintenance and for consumers of the library.
(Same comment for other new methods/constructors/classes.)

File: src/main/java/org/mozilla/zest/core/v1/ZestCookie.java [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+
+    public Date getExpiryDate() {
+        return expiry;
+    }
+
+    public void setExpiryDate(Date expiry) {
+        this.expiry = expiry;
+    }
+
+    public boolean getSecure() {
```
Comment (thc202) [inline src/main/java/org/mozilla/zest/core/v1/ZestCookie.java]: `isSecure` instead.

File: src/main/java/org/mozilla/zest/core/v1/ZestRequest.java [hunk 3]
```diff
@@ -77,11 +78,11 @@ public ZestRequest deepCopy () {
 			zr.addAssertion((ZestAssertion)zt.deepCopy());
 		}
 		
-		for (Cookie cookie : this.cookies) {
-			zr.addCookie(new Cookie(cookie.getDomain(), cookie.getName(), cookie.getValue(),
+		for (ZestCookie cookie : this.cookies) {
+			zr.addCookie(new ZestCookie(cookie.getDomain(), cookie.getName(), cookie.getValue(),
 					cookie.getPath(), cookie.getExpiryDate(), cookie.getSecure()));
 		}
-		zr.cookies = this.cookies;
```
Comment (thc202) [inline src/main/java/org/mozilla/zest/core/v1/ZestRequest.java]: - This warrants a test, hopefully the current code is not relying on this bug. 
 - Ideally these behaviour changes should be done in another changeset than a refactor (specially if not noted anywhere), makes it easer to review.

File: src/main/java/org/mozilla/zest/core/v1/ZestRequest.java [hunk 4]
```diff
@@ -299,6 +300,14 @@ public void replaceTokens(ZestVariables tokens) {
 		this.setMethod(tokens.replaceInString(this.getMethod(), false));
 		this.setHeaders(tokens.replaceInString(this.getHeaders(), false));
 		this.setData(tokens.replaceInString(this.getData(), false));
+
+		List<ZestCookie> replacedCookies = new ArrayList<>();
+		for (ZestCookie cookie : this.cookies) {
+			replacedCookies.add(new ZestCookie(cookie.getDomain(), tokens.replaceInString(cookie.getName(), false),
```
Comment (thc202) [inline src/main/java/org/mozilla/zest/core/v1/ZestRequest.java]: - How about replace the other fields (i.e. domain, path)? They can also have variables.
 - Does it really need to create new cookies? (It could use the setters.)

File: src/main/java/org/mozilla/zest/core/v1/ZestRequest.java [hunk 5]
```diff
@@ -326,15 +335,31 @@ public void clearCookies() {
...
-		this.addCookie(new Cookie(domain, name, value, path, expiry, secure));
+		this.addCookie(new ZestCookie(domain, name, value, path, expiry, secure));
 	}
 	
-	public void addCookie(Cookie cookie) {
+	public void addCookie(ZestCookie cookie) {
 		this.cookies.add(cookie);
 	}
 
+	@Obsolete
```
Comment (thc202) [inline src/main/java/org/mozilla/zest/core/v1/ZestRequest.java]: - Use `Deprecated` instead (same for following method).
 - Worth adding the JavaDoc pointing to the new method, also worth adding the version in which was deprecated, e.g.:
```Java
/**
 * @param cookie the cookie to add.
 * @deprecated (0.14) Use {@link #addCookie(ZestCookie)} instead.
 */
@Deprecated
public void addCookie(Cookie cookie) {
```
(Same comment for the other deprecated methods/constructors.)

File: src/main/java/org/mozilla/zest/core/v1/ZestRequest.java [hunk 6]
```diff
@@ -326,15 +335,31 @@ public void clearCookies() {
...
+	public void addCookie(Cookie cookie) {
+		this.cookies.add(new ZestCookie(cookie.getDomain(), cookie.getName(), cookie.getValue(),
```
Comment (thc202) [inline src/main/java/org/mozilla/zest/core/v1/ZestRequest.java]: Could call the method `addCookie(String domain,....)` instead.

File: src/main/java/org/mozilla/zest/impl/HttpClientWrapper.java [hunk 7]
```diff
@@ -0,0 +1,169 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+package org.mozilla.zest.impl;
+
+import org.apache.commons.httpclient.*;
```
Comment (thc202) [inline src/main/java/org/mozilla/zest/impl/HttpClientWrapper.java]: Use specific imports, easier to check which classes are in use, also, star imports are not used throughout the codebase.

File: src/main/java/org/mozilla/zest/impl/HttpClientWrapper.java [hunk 8]
```diff
@@ -0,0 +1,169 @@
...
+import org.mozilla.zest.core.v1.ZestAuthentication;
+import org.mozilla.zest.core.v1.ZestHttpAuthentication;
+import org.mozilla.zest.core.v1.ZestRequest;
+import org.mozilla.zest.core.v1.ZestResponse;
+
+import java.io.IOException;
+import java.util.Date;
+import java.util.Locale;
+
+public class HttpClientWrapper implements ZestHttpClient{
```
Comment (thc202) [inline src/main/java/org/mozilla/zest/impl/HttpClientWrapper.java]: - Does it need to be public? Are there plans for other "external" classes to use/access it?
 - Could be called `CommonsHttpClient` instead, it's an implementation.

File: src/main/java/org/mozilla/zest/impl/HttpClientWrapper.java [hunk 9]
```diff
@@ -0,0 +1,169 @@
...
+            setHttpAuthentication((ZestHttpAuthentication) zestAuthentication);
+        }
+    }
+
+    @Override
+    public void setProxy(String host, int port) {
+        httpclient.getHostConfiguration().setProxy(host, port);
+    }
+
+    public void setHttpAuthentication(ZestHttpAuthentication zestHttpAuthentication) {
```
Comment (thc202) [inline src/main/java/org/mozilla/zest/impl/HttpClientWrapper.java]: Private? Also, it could be inlined.

File: src/main/java/org/mozilla/zest/impl/ZestBasicRunner.java [hunk 10]
```diff
@@ -63,10 +63,11 @@
 import org.mozilla.zest.core.v1.ZestVariables;
 import org.openqa.selenium.WebDriver;
 
+
```
Comment (thc202) [inline src/main/java/org/mozilla/zest/impl/ZestBasicRunner.java]: Should be reverted.

File: src/main/java/org/mozilla/zest/impl/ZestBasicRunner.java [hunk 11]
```diff
@@ -450,130 +458,6 @@ public String runScript(String script, Map<String, String> params)
...
-		
-		
-		if (method.getStatusCode() == 302 && req.isFollowRedirects() && ! req.getMethod().equals("GET")) {
-			// Follow the redirect 'manually' as the httpclient lib only supports them for GET requests
-			method = new GetMethod(method.getResponseHeader("Location").getValue());
-			// Just in case there are multiple redirects
-			method.setFollowRedirects(req.isFollowRedirects());
-
-			try {
-				this.debug(req.getMethod() + " : " + req.getUrl());
```
Comment (thc202) [inline src/main/java/org/mozilla/zest/impl/ZestBasicRunner.java]: We lost this debug statement :)

File: src/main/java/org/mozilla/zest/impl/ZestHttpClient.java [hunk 12]
```diff
@@ -0,0 +1,19 @@
...
+
+import org.mozilla.zest.core.v1.ZestAuthentication;
+import org.mozilla.zest.core.v1.ZestRequest;
+import org.mozilla.zest.core.v1.ZestResponse;
+
+import java.io.IOException;
+
+public interface ZestHttpClient{
+
+    void setAuthentication(ZestAuthentication zestAuthentication);
```
Comment (thc202) [inline src/main/java/org/mozilla/zest/impl/ZestHttpClient.java]: - Maybe `addAuthentication`? "set" implies the previous authentication is replaced, while it might not be the case (it's cumulative if targeting different hosts).
 - Any special reason to use `ZestAuthentication` instead of `ZestHttpAuthentication`?

File: src/test/java/org/mozilla/zest/test/v1/ZestRequestUnitTest.java [hunk 13]
```diff
@@ -44,4 +46,31 @@ public void testTokenReplacement() throws Exception {
...
+		tokens.addVariable("token", "ABC");
+		tokens.addVariable("token1", "DEFG");
+		tokens.addVariable("token2", "GHI");
+		tokens.addVariable("token3", "JKL");
+		req2.replaceTokens(tokens);
+
+		assertEquals ("http://www.example.com/app/DEFG", req2.getUrl().toString());
+		assertEquals ("Set-Cookie: test=GHI", req2.getHeaders());
+		assertEquals ("test=JKL&user=12JKL34", req2.getData());
+		assertEquals ("12JKL34", req2.getZestCookies().get(0).getName());
```
Comment (thc202) [inline src/test/java/org/mozilla/zest/test/v1/ZestRequestUnitTest.java]: This could be done in the previous method, or, if you really want a different method the other assertions should be removed (no need to duplicate them).

File: src/main/java/org/mozilla/zest/impl/HttpClientWrapper.java [hunk 14]
```diff
@@ -0,0 +1,169 @@
...
+                method = new TraceMethod(uri.toString());
+                break;
+            default:
+                throw new IllegalArgumentException("Method not supported: "
+                        + req.getMethod());
+        }
+
+        setHeaders(method, req.getHeaders());
+
+        for (Cookie cookie : req.getCookies()) {
```
Comment (thc202) [inline src/main/java/org/mozilla/zest/impl/HttpClientWrapper.java]: Probably better to use `getZestCookies()`, while it's easier to use `getCookies()` it will be removed sooner or later.

File: src/main/java/org/mozilla/zest/core/v1/ZestCookie.java (hunk 1 above)
Comment (denniskniep) [reply src/main/java/org/mozilla/zest/core/v1/ZestCookie.java]: Okay I do my best =)

...but I dont know if life is easier if an interface `ZestRunner` has the comment "The Interface ZestRunner." ;)

Comment (thc202) [reply src/main/java/org/mozilla/zest/core/v1/ZestCookie.java]: Note that I'm not asking you to document it, just add the since tag.

File: src/main/java/org/mozilla/zest/impl/HttpClientWrapper.java (hunk 8 above)
Comment (denniskniep) [reply src/main/java/org/mozilla/zest/impl/HttpClientWrapper.java]: #Does it need to be public? 
I think so, because if someone wants to change the behavior with `setParams()` or `setAuthentication()`  he must create an instance of this class and inject it in `ZestBasicRunner`

Comment (thc202) [reply src/main/java/org/mozilla/zest/impl/HttpClientWrapper.java]: They wouldn't need direct access to this class, they can use the existing methods (e.g. `setHttpClientParams`), maybe it would be easier to reuse this one if other changes are needed but the idea is to hide the use of Commons HttpClient to be able remove it sooner rather than later(?). One can always set a custom `ZestHttpClient` implementation if really needs to change how the messages are sent (or, the interface should be changed to expose that functionality if deemed general/useful for the majority?)

File: src/main/java/org/mozilla/zest/impl/ZestHttpClient.java (hunk 12 above)
Comment (denniskniep) [reply src/main/java/org/mozilla/zest/impl/ZestHttpClient.java]: #Any special reason to use ZestAuthentication instead of ZestHttpAuthentication?
For the case that there is another Authentication i.e. jwt bearer token in the header

Comment (thc202) [reply src/main/java/org/mozilla/zest/impl/ZestHttpClient.java]: OK, that's what I suspected, but more work would be required to support other authentication methods ;)

File: src/main/java/org/mozilla/zest/impl/HttpClientWrapper.java (hunk 8 above)
Comment (denniskniep) [reply src/main/java/org/mozilla/zest/impl/HttpClientWrapper.java]: > they can use the existing methods (e.g. setHttpClientParams, setHttpClient)
okay. But then we should not deprecate setHttpClientParams and setHttpClient until there is a replacement for Commons HttpClient, should't we?

Comment (thc202) [reply src/main/java/org/mozilla/zest/impl/HttpClientWrapper.java]: I think it's fine to deprecate those methods, it's now possible to set a custom `ZestHttpClient` (in case of ZAP it would just "redirect" to existing classes).

File: src/main/java/org/mozilla/zest/core/v1/ZestOutputWriter.java [hunk 15]
```diff
@@ -0,0 +1,14 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+package org.mozilla.zest.core.v1;
+
+/**
+ * @since 0.14
+ */
+public interface ZestOutputWriter {
```
Comment (thc202) [inline src/main/java/org/mozilla/zest/core/v1/ZestOutputWriter.java]: Don't think we need to introduce this class just because of that debug statement (I was just kidding with the lost debug statement), but it might be useful in other places? Maybe `ZestRuntime` should extend it (that's the interface that already provides these methods)? Instead of `ZestBasicRunner` implement it.

File: src/main/java/org/mozilla/zest/core/v1/ZestRequest.java [hunk 16]
```diff
@@ -10,6 +10,7 @@
 import java.util.List;
 
 import org.apache.commons.httpclient.Cookie;
+import org.apache.http.annotation.Obsolete;
```
Comment (thc202) [inline src/main/java/org/mozilla/zest/core/v1/ZestRequest.java]: This can now be removed.

File: src/test/java/org/mozilla/zest/test/v1/ZestRequestUnitTest.java [hunk 17]
```diff
@@ -42,6 +46,39 @@ public void testTokenReplacement() throws Exception {
...
+
+		ZestRequest req2 = req.deepCopy();
+
+		assertTrue(req.getHeaders().equals(req2.getHeaders()));
+		assertTrue(req.getData().equals(req2.getData()));
+
+		ZestCookie cookie = req.getZestCookies().get(0);
+		ZestCookie cookie2 = req2.getZestCookies().get(0);
+
+		assertTrue("cookies should not be reference equals", !cookie.equals(cookie2));
```
Comment (thc202) [inline src/test/java/org/mozilla/zest/test/v1/ZestRequestUnitTest.java]: Probably better to already use `!=`, otherwise it will start to fail (while it shouldn't) when `ZestCookie` properly implements `equals`.

**Comment by denniskniep** [comment]
@thc202 I squashed the commits into one refactoring commit and one bugfix commit.

**Comment by thc202** [comment]
Thank you!
